homeassistant/components/life360/__init__.py

```diff
@@ -28,184 +22,174 @@
 
+async def async_migrate_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Migrate config entry."""
+    LOGGER.debug("Migrating config entry from version %s", entry.version)
 
-def setup(hass: HomeAssistant, config: ConfigType) -> bool:
-    """Set up integration."""
-    conf = config.get(DOMAIN, LIFE360_SCHEMA({}))
-    hass.data[DOMAIN] = {"config": conf, "apis": {}}
-    discovery.load_platform(hass, Platform.DEVICE_TRACKER, DOMAIN, None, config)
-
-    if CONF_ACCOUNTS not in conf:
-        return True
-
-    # Check existing config entries. For any that correspond to an entry in
-    # configuration.yaml, and whose password has not changed, nothing needs to
-    # be done with that config entry or that account from configuration.yaml.
-    # But if the config entry was created by import and the account no longer
-    # exists in configuration.yaml, or if the password has changed, then delete
-    # that out-of-date config entry.
-    already_configured = []
-    for entry in hass.config_entries.async_entries(DOMAIN):
-        # Find corresponding configuration.yaml entry and its password.
-        password = None
-        for account in conf[CONF_ACCOUNTS]:
-            if account[CONF_USERNAME] == entry.data[CONF_USERNAME]:
-                password = account[CONF_PASSWORD]
-        if password == entry.data[CONF_PASSWORD]:
-            already_configured.append(entry.data[CONF_USERNAME])
-            continue
-        if (
-            not password
-            and entry.source == config_entries.SOURCE_IMPORT
-            or password
-            and password != entry.data[CONF_PASSWORD]
-        ):
-            hass.async_create_task(hass.config_entries.async_remove(entry.entry_id))
-
-    # Create config entries for accounts listed in configuration.
-    for account in conf[CONF_ACCOUNTS]:
-        if account[CONF_USERNAME] not in already_configured:
-            hass.async_create_task(
-                hass.config_entries.flow.async_init(
-                    DOMAIN,
-                    context={"source": config_entries.SOURCE_IMPORT},
-                    data=account,
-                )
-            )
+    if entry.version == 1:
+        entry.version = 2
+        entry.source = SOURCE_USER
+        unique_id = entry.data[CONF_USERNAME].lower()
+        hass.config_entries.async_update_entry(
+            entry,
+            unique_id=unique_id,
+            title=unique_id,
+            options=hass.data[DOMAIN]["cfg_options"],
+        )
+
+    LOGGER.info("Config entry migration to version %s successful", entry.version)
     return True
 
 
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up config entry."""
-    hass.data[DOMAIN]["apis"][entry.data[CONF_USERNAME]] = get_api(
-        entry.data[CONF_AUTHORIZATION]
+    account = hass.data[DOMAIN]["accounts"].setdefault(
+        cast(str, entry.unique_id), AccountData()
```

Comment:
Always use the config entry entry_id for storing in hass.data.

Reply:
Can't really do that since at least in one place I'm storing data before the config entry has been created. See line 133 in config_flow.py (right before returning async_create_entry.)

Reply:
We should not store the api in hass.data in the config flow.

Reply:
Ok